File: ZOSInteractServer/ZOSInteractServer/Resource.cpp
#include "rest_framework.h"

BaseResource::BaseResource(const std::string &p_path) : path(p_path)
{
    //
}

Resource::Resource(const std::string &p_path, const Method& verb, 
    resource_callback callback) 
    : BaseResource(p_path), m_method(verb), m_callback(callback)
{
    //
}

Response Resource::handle_request(const Request &request) const
{
    return m_callback(request);
}

WebSocketResource::WebSocketResource(const std::string &p_path, WebSocketResourceCallbacks callbacks) : BaseResource(path), m_callbacks(callbacks)
{
    //
}

File: rust_http_server/http/resource.rs
use std::sync::Arc;

use http::method::Method;
use http::request::Request; 
use http::response::Response;

//type RequestCallback<'a> = &'a Fn((Request, Response));

pub struct Resource {
	pub path: String,
	pub method: Method,
	request_callback: Box<Fn(&Request, &mut Response) + Send + Sync + 'static>
}

impl Resource {
	pub fn init(path: &str, method: Method, 
		callback: Box<Fn(&Request, &mut Response) + Send + Sync + 'static>) -> Resource
	{
		Resource { path: path.to_string(), method: method, request_callback: callback}
	}
	pub fn handle_request(&self, request: &Request, response: &mut Response) {
		(*self.request_callback)(request, response);
	}
}